src/GafferImage/AddDepth.cpp

```diff
@@ -0,0 +1,316 @@
+	{
+		result.push_back( "Z" );
+	}
+
+	if( find( result.begin(), result.end(), "ZBack" ) == result.end() )
+	{
+		result.push_back( "ZBack" );
+	}
+
+	return resultData;
+}
+
+void AddDepth::hashChannelData( const GafferImage::ImagePlug *output, const Gaffer::Context *context, IECore::MurmurHash &h ) const
+{
+	FlatImageProcessor::hashChannelData( output, context, h );
+	const std::string channelName = context->get<std::string>( ImagePlug::channelNameContextName );
+
+	const std::string sourceZChannel = sourceZChannelPlug()->getValue();
+	const std::string sourceZBackChannel = sourceZBackChannelPlug()->getValue();
+
+	if ( channelName == "Z" )
+	{
+		if( !sourceZChannel.size() )
+		{
+			depthPlug()->hash( h );
+		}
+		else
+		{
+			ConstStringVectorDataPtr channelNamesData = inPlug()->channelNamesPlug()->getValue();
+			const vector<string> &channelNames = channelNamesData->readable();
+
+			if( find( channelNames.begin(), channelNames.end(), sourceZChannel ) == channelNames.end() )
+			{
+				throw( IECore::Exception( "AddDepth : Cannot find requested Z channel - no channel \""
+					+ sourceZChannel + "\" found."
+				) );
+			}
+
+			ImagePlug::ChannelDataScope zScope( context );
+			zScope.setChannelName( sourceZChannel );
+			h = inPlug()->channelDataPlug()->hash();
+		}
+	}
+	else if ( channelName == "ZBack" )
+	{
+		if( !sourceZChannel.size() && !sourceZBackChannel.size() )
+		{
+			depthPlug()->hash( h );
+			thicknessPlug()->hash( h );
+		}
+		else
+		{
+			ConstStringVectorDataPtr channelNamesData = inPlug()->channelNamesPlug()->getValue();
+			const vector<string> &channelNames = channelNamesData->readable();
+
+			if( sourceZBackChannel.size() )
+			{
+				if( find( channelNames.begin(), channelNames.end(), sourceZBackChannel ) == channelNames.end() )
+				{
+					throw( IECore::Exception( "AddDepth : Cannot find requested ZBack channel - no channel \""
+						+ sourceZBackChannel + "\" found."
+					) );
+				}
+				ImagePlug::ChannelDataScope zBackScope( context );
+				zBackScope.setChannelName( sourceZBackChannel );
+				h = inPlug()->channelDataPlug()->hash();
+			}
+			else
+			{
+				thicknessPlug()->hash( h );
+
+				ImagePlug::ChannelDataScope zScope( context );
+				zScope.setChannelName( sourceZChannel );
+				outPlug()->channelDataPlug()->hash(h);
+			}
+		}
+	}
+	else
+	{
+		h = inPlug()->channelDataPlug()->hash();
+	}
+}
+
+IECore::ConstFloatVectorDataPtr AddDepth::computeChannelData( const std::string &channelName, const Imath::V2i &tileOrigin, const Gaffer::Context *context, const ImagePlug *parent ) const
+{
+	const std::string sourceZChannel = sourceZChannelPlug()->getValue();
+	const std::string sourceZBackChannel = sourceZBackChannelPlug()->getValue();
+
+	if ( channelName == "Z" )
+	{
+		if( !sourceZChannel.size() )
+		{
+			// Set constant Z
+			float depth = depthPlug()->getValue();
+			FloatVectorDataPtr resultData = new FloatVectorData;
+			resultData->writable().resize( ImagePlug::tilePixels(), depth );
```

Comment:
Just checked, and as per our discussion on hangouts, I had stuffed up my "Text and AddDepth" test by starting it with a Constant with alpha 1. Giving the background an alpha of 0 makes that all work fine, making the Deep
Tidy unnecessary prior to the DeepHoldout.

But there's still a problem whereby DeepTidy claims to output a flat image after pruning transparent samples from a flat input. This confuses downstream nodes like DeepSampler, which then enter code paths specific to images with one-sample-per-pixel.

Reply:
I want to be exactly clear here:
> But there's still a problem whereby DeepTidy claims to output a flat image after pruning transparent samples from a flat input.
This makes it sound like DeepTidy is outputting inconsistent output.
What is actually happening could be more accurately described as:
"DeepTidy notices the input is flat, assumes Flat is better than Tidy, and so passes through the input completely unchanged".

I can see where you're coming from with the argument that tidying with prune on a flat image should unflatten, but there are probably arguments either way.  It's not like there are many circumstances where this would be useful - any Gaffer node that takes a deep image can take a flat image, and doing extra processing beforehand is quite unlikely to be a performance benefit.